lib/Backends/NNPI/NNPI.cpp

```diff
@@ -1313,3 +1331,260 @@ Error NNPIBackend::bindContexts(
+    break;
+  case glow::ElemKind::UInt8QTy:
+    desc.quantParams.precision = NNPI_PRECISION_UINT8;
+    desc.quantParams.type = NNPI_QUANTIZATION_GEMMLOWP;
+    break;
+  case glow::ElemKind::Int32ITy:
+    desc.quantParams.precision =
+        NNPI_PRECISION_INT32; // The backend will convert to Int32 when
+                              // compiling.
+    desc.quantParams.type = NNPI_QUANTIZATION_NONE;
+    break;
+  case glow::ElemKind::Int64ITy:
+    desc.quantParams.precision =
+        NNPI_PRECISION_INT32; // The backend will convert to Int32 when
+                              // compiling.
+    desc.quantParams.type = NNPI_QUANTIZATION_NONE;
+    break;
+  case glow::ElemKind::Int32QTy:
+    desc.quantParams.precision = NNPI_PRECISION_INT32;
+    desc.quantParams.type = NNPI_QUANTIZATION_GEMMLOWP;
+    break;
+  case glow::ElemKind::UInt8FusedQTy:
+    desc.quantParams.precision = NNPI_PRECISION_UINT8;
+    desc.quantParams.type = NNPI_QUANTIZATION_GEMMLOWP_PCQ_FUSED;
+    break;
+  case glow::ElemKind::UInt8FusedFP16QTy:
+    desc.quantParams.precision = NNPI_PRECISION_UINT8;
+    desc.quantParams.type = NNPI_QUANTIZATION_GEMMLOWP_PCQ_FUSED_FP16;
+    break;
+  case glow::ElemKind::UInt4FusedFP16QTy:
+    desc.quantParams.precision = NNPI_PRECISION_UINT8;
+    desc.quantParams.type = NNPI_QUANTIZATION_GEMMLOWP_PCQ_4BIT_FUSED_FP16;
+    break;
+  case glow::ElemKind::BoolTy:
+    desc.quantParams.precision = NNPI_PRECISION_BOOLEAN;
+    desc.quantParams.type = NNPI_QUANTIZATION_NONE;
+    break;
+
+  default:
+    LOG_AND_RETURN_IF(ERROR, true, "Invalid type", false);
+    break;
+  }
+  memset(&desc.quantParams.params, 0,
+         sizeof(desc.quantParams.params)); // Actual values are not needed here.
+
+  desc.attributes.value = 0; // No attributes needed here.
+
+  return true;
+}
+
+// Prepare the list of NNPITensorDesc for the estimate call.
+static bool UpdateDescListForEstimate(std::vector<NNPITensorDesc> &descs,
+                                      std::vector<glow::TypeRef> types) {
+  if (descs.size() != types.size()) {
+    return false;
+  }
+  bool retVal = true;
+  for (size_t i = 0; i < descs.size(); i++) {
+    if (types.at(i) != nullptr) {
+      retVal &= UpdateDescForEstimate(descs.at(i), types.at(i));
+    }
+  }
+  return retVal;
+}
+
+double NNPIBackend::estimateEmbeddingNode(const glow::NodeInfo &NI,
+                                          bool fp32Accumulation,
+                                          glow::LengthsMode lengthsMode,
+                                          float averageLength) const {
+  if (!isOpSupported(NI)) {
+    // Op isn't supported.
+    return -1.0;
+  }
+  NNPI_LENGTH_TYPE lengthType = NNPI_LENGTH_VARIABLE;
+  LOG_AND_RETURN_IF(ERROR,
+                    NNPIImporter::convertLengthsModeToLengthType(
+                        lengthsMode, lengthType) != NNPI_NO_ERROR,
+                    "Failed to convert LengthsMode", -1.0);
+
+  enum DescIndex {
```

Comment:
Unused?

Reply:
Implicitly used in the descs vector.
i.e. the first one is for input, the second is for output etc.

Reply:
Maybe makes sense to move it down to just before `nnpiEstimateSparseLengthsWeightedSumOp()`. Or even just remove it and use `/* Input */ 0` in the function call. No biggy though.

Reply:
It is used see line #1581.